pandas/plotting/_matplotlib/core.py

```diff
@@ -240,18 +246,30 @@ def _iter_data(self, data=None, keep_index=False, fillna=None):
         if fillna is not None:
             data = data.fillna(fillna)
 
-        for col, values in data.items():
-            if keep_index is True:
-                yield col, values
-            else:
-                yield col, values.values
+        if self.by is None:
+            for col, values in data.items():
+                if keep_index is True:
+                    yield col, values
+                else:
+                    yield col, values.values
+        else:
+            cols = data.columns.get_level_values(0).unique()
```

Comment:
So we only plot the first level of a multi index then? Is this limitation documented anywhere?

Reply:
emm, I would not call it a limitation, this branch will not be reached even dataset is a MI dataframe while `by` is not defined (`None`). The reason to have it is to separate the groups caused by `by` argument, and produce iterated data for `plt` to plot, all levels in MI will be plotted, could check it in the tests.

Reply:
OK I think I understand. Doesn't this require `by` and the first level of the columns to be the same though?

I think this can be more cleary written if you groupby the `by` argument. If you do that you can then put the `for col, val in data.items()` loop into a function and just apply that function along the groupby object

Reply:
@WillAyd thanks very much for your comment, appreciate a lot!

this data here in `_iter_data` has been grouped, could reference to the patch i add here in `compute_plot_data`:
```python
# GH15079 restructure data if by is defined
        if self.by is not None:
            self.subplots = True
            grouped = data.groupby(self.by)

            data_list = []
            for key, group in grouped:
                columns = MultiIndex.from_product([[key], self.column])
                sub_group = group[self.column]
                sub_group.columns = columns
                data_list.append(sub_group)

            data = concat(data_list, axis=1)
```
And here, in `_iter_data`, we need to sub-select columns based on the first level of MI, i simplified this block in `_iter_data` again to avoid repeatedly use the for loop, and add comment to clarify it, feel free to give another look! thanks!